python/pyxet/pyxet/sync.py

```diff
@@ -0,0 +1,221 @@
+from abc import ABC, abstractmethod
+from concurrent.futures import ThreadPoolExecutor
+from datetime import datetime
+from functools import partial
+
+from pyxet.util import _ltrim_match, _get_fs_and_path, _single_file_copy, _isdir
+
+XET_MTIME_FORMAT = '%Y-%m-%dT%H:%M:%S%z'
+
+
+class SyncCommand:
+
+    def __init__(self, source, destination, use_mtime, message, dryrun):
+        self._message = message
+        self._dryrun = dryrun
+        self._src_fs, self._src_proto, self._src_root = _get_normalized_fs_protocol_and_path(source)
+        self._dest_fs, self._dest_proto, self._dest_root = _get_normalized_fs_protocol_and_path(destination)
+        self._use_mtime = use_mtime
+        if use_mtime:
+            self._cmp = MTimeSyncComparator(self._src_proto, self._dest_proto)
+        else:
+            self._cmp = SizeOnlySyncComparator()
+
+    def validate(self):
+        """
+        Performs early validation for the source and destination paths of
+        a xet sync. Doesn't catch all failure conditions, but catches many
+        of the easy validations.
+
+        Raises exceptions on failure
+        """
+
+        print(f"Checking sync")
+        if self._dest_proto != 'xet':
+            raise ValueError(f"Unsupported destination protocol: {self._dest_proto}, only xet:// targets are supported")
+        if self._src_proto == 'xet':
+            raise ValueError(f"Unsupported source protocol: {self._src_proto}, only non-xet sources are supported")
+
+        # check that the destination specifies an existing branch
+        # TODO: we may want to be able to sync remote location to a new branch?
+        self._dest_fs.branch_info(self._dest_root)
+
+        # source should be a directory
+        if not _isdir(self._src_fs, self._src_root):
+            raise ValueError(f"source: {self._src_root} needs to be a directory")
+
+        # wildcards not supported
+        if '*' in self._src_root or '*' in self._dest_root:
+            raise ValueError(f"Wildcards not supported in paths")
+
+    def run(self):
+        """
+        Runs this Sync command.
+        """
+        print(f"Starting sync")
+        failed_copies = []
+
+        if not self._dryrun:
+            self._dest_fs.start_transaction(self._message)
+        with ThreadPoolExecutor() as executor:
+            futures = []
+            if self._use_mtime:
+                self._sync_with_info(executor, futures, self._src_root, self._dest_root)
+            else:
+                self._sync_with_ls(executor, futures, self._src_root, self._dest_root)
+
+            # Waiting for all copy jobs to complete
+            files_synced = 0
+            files_ignored = 0
+            for future in futures:
+                try:
+                    i = future.result()
```

Comment:
Bit confused with `i<0` check, since I think you are returning a bool. Might be code drift or worth commenting